### PR TITLE
librdmacm: Add support for extended join multicast API

### PR DESCRIPTION
**Pull request by alexvesker**
Add rdma_join_multicast_ex. This extended API can be extended
to support more features.

Currently, support for specifying MC join flags was added,
the following Multicast join flags are supported:

-Full Member:
 The initiator creates the Multicast group(MCG) if it wasn't
 previously created, can send Multicast messages to the group
 and receive messages from the MCG.

-Send Only Full Member:
 The initiator creates the Multicast group(MCG) if it wasn't
 previously created, can send Multicast messages but doesn't
 receive any messages from the MCG (send-only).

 librdmacm previously supported only "Full Member" join.

Signed-off-by: Alex Vesker valex@mellanox.com
Reviewed by: Hal Rosenstock hal@mellanox.com

### CODE REVIEW

File: include/rdma/rdma_cma.h [hunk 1]
```diff
@@ -197,6 +197,29 @@ struct rdma_addrinfo {
...
+	RDMA_CM_JOIN_MC_ATTR_JOIN_FLAGS	= 1 << 1,
+	RDMA_CM_JOIN_MC_ATTR_RESERVED	= 1 << 2,
+};
+
+/* Multicast join flags */
+enum rdma_cm_mc_join_flags {
+        RDMA_MC_JOIN_FLAG_FULLMEMBER,
+        RDMA_MC_JOIN_FLAG_SENDONLY_FULLMEMBER,
+        RDMA_MC_JOIN_FLAG_RESERVED,
+};
```
Comment (shefty) [inline include/rdma/rdma_cma.h:212]: Flags should not be in a named enum, as the enum itself cannot be used.  And flags should be bit values.

I prefer to keep this simple and just add an RDMA_CM_SENDONLY flag.

File: include/rdma/rdma_cma.h [hunk 2]
```diff
@@ -538,6 +561,30 @@ int rdma_join_multicast(struct rdma_cm_id *id, struct sockaddr *addr,
...
+ *   multicast address to an RDMA device.  The user must call
+ *   rdma_leave_multicast to leave the multicast group and release any
+ *   multicast resources.  The context is returned to the user through
+ *   the private_data field in the rdma_cm_event.
+ * See also:
+ *   rdma_leave_multicast, rdma_bind_addr, rdma_resolve_addr, rdma_create_qp
+ */
+int rdma_join_multicast_ex(struct rdma_cm_id *id,
+			   struct rdma_cm_join_mc_attr_ex *mc_join_attr,
+			   void *context);
```
Comment (shefty) [inline include/rdma/rdma_cma.h:585]: I'm torn between trying to make this extensible, versus just adding a flag parameter.

Comment (alexvesker) [reply include/rdma/rdma_cma.h:585]: The flag-based approach doesn't solve one of the main concerns
which we have with current code - lack of extensibility.
Right now, we are proposing new addition which solves it and
removes future compatibility issues if any.

Let's take an example - rdma_join_multicat().
This function was written without taking into consideration future
extensions and no one had thoughts at that time that multicast join 
send only is needed.

File: include/rdma/rdma_cma.h (hunk 1 above)
Comment (alexvesker) [reply include/rdma/rdma_cma.h:212]: Will remove the name from these enums. The join flags
are corresponding to the join flags in the kernel under
rdma_user_cm.h "Multicast join flags", this will require a
kernel change. Regarding the bit value unlike the comp mask
which should be orthogonal it is not clear why it should be changed.

Comment (shefty) [reply include/rdma/rdma_cma.h:212]: The rdma cm is intended to mimic a sockets-like API, not a native IB one.  The current code matches with setsockopt - IP_ADD_MEMBERSHIP.  There's no mapping for 'full member' for sockets.

Unlike sockets, we do have to join for send only transfers.  For that I'm suggesting that we add a flag, SENDONLY, that the app can set when joining.  If left unset, the QP is attached for send/receive.  This will also allow us to add other flags, like RDMA_MC_NO_LOOPBACK, to disable loopback traffic.

We don't need a kernel change.  We should be able to map what the user provides to whatever the kernel accepts.

Basically, I'm trying to keep the interface simpler to use/understand.  A SENDONLY flag clearly specifies what it does.  Saying 'FULLMEMBER' requires someone to go read a 1500 page spec and figure out what exactly that means on IB.  And I have no idea what that would mean for UDP if iWarp adapters ever decide to support UDP.

File: include/rdma/rdma_cma.h (hunk 2 above)
Comment (shefty) [reply include/rdma/rdma_cma.h:585]: We have some options here.  We could use rdma_set_option to specify more detailed/different multicast join operations, similar to setsockopt.  We could also define a new sockaddr type that allows us to specify more details -- this is more of a hack, but would be compatible.

If we have a new API, as long as we have a void \* parameter, we can use a flag to indicate/change how that parameter should be interpreted by the library.  So there we can be extensible by adding a new flag parameter.

There's a trade off between being extensible and easy to use here.  Currently, it looks like the only goal is to support send-only joins.

As an option, we could define this:

rdma_join_multicast(struct rdma_cm_id *id, void *mc_addr, void *context, uint64_t flags);

Then the address doesn't need to be a sockaddr either, which would allow specifying the full set of IB multicast data.

Comment (alexvesker) [reply include/rdma/rdma_cma.h:585]: Ok I understand your approach, I agree it can make things easier for the users. 
Regarding your example, using the same function name and adding a variable can break current API so I will have to add a new function, something like rdma_join_multicast_flags(...).

Comment (shefty) [reply include/rdma/rdma_cma.h:585]: I think your function name (add _ex) is fine.  I was just being terse in my responses.  Thanks

File: include/rdma/rdma_cma.h (hunk 1 above)
Comment (alexvesker) [reply include/rdma/rdma_cma.h:212]: After some thinking and reading your comments (more than once), 
the main problem I see is that SENDONLY != SENDONLY_FULLMEMBER.
I implemented kernel code support for SENDONLY_FULLMEMBER,
using a flag called SENDONLY can be misleading since it might
also be a future valid option (already in IB spec.).

Regarding the bit values, currently we provide in the command struct a user value for the 
join state, the kernel translates it to the kernel supported value in ucma based on enums/defines in  
rdma_user_cm.h this means that both should be aligned. To use bit values we will need to:
Modify rdma_user_cm.h in the kernel or add a redundant enum that translates the value in librdmacm.

The solution below is a combination of your suggestion with minor changes:
Prototype: rdma_join_multicast_ex(.., join_flags)
Flags: FULLMEMBER or SENDONLY_FULLMEMBER which are bit value
The flags above will be translated based on the values in rdma_user_cm.h (not bit value)
which means we don't need to change the kernel and wait another cycle.

**Comment by shefty** [comment]
> After some thinking and reading your comments (more than once),
> the main problem I see is that SENDONLY != SENDONLY_FULLMEMBER.
> I implemented kernel code support for SENDONLY_FULLMEMBER,
> using a flag called SENDONLY can be misleading since it might
> also be a future valid option (already in IB spec.).

The rdma cm is a transport neutral API.  Please keep this in mind and try to avoid exposing transport specific limitations/definitions up through this API.

> The solution below is a combination of your suggestion with minor
> changes:
> Prototype: rdma_join_multicast_ex(.., join_flags)
> Flags: FULLMEMBER or SENDONLY_FULLMEMBER which are bit value
> The flags above will be translated based on the values in
> rdma_user_cm.h (bit value)
> which means we don't need to change the kernel and wait another cycle.

The issue is that send-only should be a stand-alone flag.  Either the app only wants to send or they want to send and receive.  Full membership should be a separate flag.  AFAICT, the MCMemberRecord flags are defined the way they are for atomicity purposes, so that a port can be downgraded from a FullMember to a NonMember without losing connectivity.  However, a software API shouldn't be designed around the wire protocol.

IMO, the question isn't around SendOnly, but whether the rdma cm should export the FullMember/NonMember concept.  IIRC, the current API assumes that all joins are FullMember -- the multicast group is automatically created if it doesn't exist.  Adding a SendOnly flag doesn't change that behavior.  So the gap in the mapping would be performing a NonMember or SendOnlyNonMember join, as that aspect of the IB architecture is currently abstracted away from the API.

Btw, please direct future PRs against this tree:

https://github.com/linux-rdma/rdma-core
- Sean
